lib/srv/transport/stream.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/gravitational/trace"
+)
+
+// Stream abstracts gRPC streams used to transport
+// frames of data.
+type Stream interface {
```

Comment:
I realize this is based off the gRPC interface, but it looks awfully similar to `io.Reader` and `io.Writer` - have you thought about reusing those interfaces here too?

Notably, the signature of `Receive` lends itself to doing lots of allocations rather than writing into an existing buffer.